Fixed lint lua errors in src

sameedali opened the pull request:
This pull request has commits which fix the various warnings and errors thrown by luacheck when run on the lua files inside the src directory.

bfredl commented on src/nvim/generators/gen_api_dispatch.lua:
```diff
@@ -135,9 +135,9 @@ for i,f in ipairs(shallowcopy(functions)) do
 -- don't expose internal attributes like "impl_name" in public metadata
-exported_attributes = {'name', 'parameters', 'return_type', 'method',
+local exported_attributes = {'name', 'parameters', 'return_type', 'method',
                        'since', 'deprecated_since'}
```
need to adjust indentation of follow-up line.

bfredl commented on src/nvim/generators/gen_api_dispatch.lua:
```diff
@@ -301,19 +305,19 @@ end
-  for i = 1, #headers do
-    if headers[i]:sub(-12) ~= '.generated.h' then
-      output:write('\n#include "nvim/'..headers[i]..'"')
+local function include_headers(arg_output, arg_headers)
```
maybe instead rename the global `output` ? At least I find no precedent for `arg_` .

sameedali replied:
I renamed it to "output_handle" is that fine? I couldn't think of a more appropriate name for the global output so didnt change it. Also named the arg_headers to "headers_to_include"

sameedali committed:
fixed indentation and renamed function arguments

justinmk commented on src/nvim/generators/gen_declarations.lua:
```diff
@@ -225,14 +224,12 @@ while init ~= nil do
-  elseif init < declendpos then
-    -- Skipping over declaration
-  elseif is_needed_file then
+  elseif (not (init < declendpos)) and is_needed_file then
```
We lost documentation here (the deleted comment). Would be better to add an inline `-- luacheck: ...` ignore.

sameedali replied:
I have updated the code to reflect these changes.

justinmk commented on src/nvim/generators/gen_unicode_tables.lua:
```diff
@@ -208,9 +208,7 @@ local build_width_table = function(ut_fp, dataprops, widthprops, widths,
-        if start >= 0 and end_ + 1 == n then
-          -- Continue with the same range.
-        else
+        if not (start >= 0 and end_ + 1 == n) then
```
same here, the old code was intentional documentation. Also when adding `-- luacheck: ignore` it should only ignore the specific luacheck code number. In this case it's 542, see http://luacheck.readthedocs.io/en/stable/warnings.html

```lua
if start >= 0 and end_ + 1 == n then  -- luacheck: ignore 542
  -- Continue with the same range.
else
```

justinmk commented on src/nvim/generators/gen_declarations.lua:
```diff
@@ -225,14 +224,14 @@ while init ~= nil do
       declline = declline - 1
     end
-  elseif init < declendpos then
+  elseif init < declendpos then -- luacheck: ignore
```
-- luacheck: ignore 542

sameedali committed:
added luacheck error numbers and fixed missing documentation

Signed-off-by: sameedali <sameed.ali.94@gmail.com>

XVilka commented:
@sameedali maybe a good time to rebase and solve conflicts?

sameedali committed:
rebased and fixed luacheck errors

Date:      Sun Apr 14 16:56:11 2019 -0400

sameedali committed:
[skip appveyor] rebased luacheck errors commits

Date:      Sun Apr 14 16:56:11 2019 -0400

sameedali committed:
fixed luacheck errors in nvim/options.lua

XVilka commented:
Apparently Travis fails to pass functional tests.

blueyed commented:
@XVilka 
Needs a rebase first - worth trying if it passed then.

@sameedali 
Do you plan to pick this up again?

justinmk commented:
I will take care of this next week.

sameedali committed:
Fixed luacheck errors in src/nvim/lua/vim.lua.
This is the last remaining lua file that wasnt linted.
Changes to be committed:
      modified:   nvim/lua/vim.lua

sameedali committed:
Fixed luacheck errors in src/nvim/lua/vim.lua.
This is the last remaining lua file that wasnt linted.
Changes to be committed:
      modified:   nvim/lua/vim.lua

blueyed commented on src/nvim/lua/vim.lua:
```diff
@@ -38,8 +38,8 @@
-  local out = vim.api.nvim_call_function('system', { cmd })
-  local err = vim.api.nvim_get_vvar('shell_error')
+  local out = vim.api.nvim_call_function('system', { cmd }) --luacheck: ignore 113
+  local err = vim.api.nvim_get_vvar('shell_error') --luacheck: ignore 113
```
We should have a `.luacheckrc` instead, with:
```lua
-- Global objects defined by the C code
read_globals = {
    "vim",
}
```

blueyed replied:
See https://github.com/awesomewm/awesome/blob/3e3c34bdd194cf75cc16a87e82b11f8df3e7c5de/.luacheckrc for more ideas.

blueyed replied:
Also please consider using `[skip appveyor]` with following commits (since linting is run on Travis).

sameedali replied:
Okay, I have added the .luacheck rc in src/ and removed the luacheck ignore flags for vim global var in this file.

blueyed replied:
For future commits, please use `[skip appveyor]` with the commit subject (first line).

sameedali commented:
Hi,
Sorry about the delay. I just rebased it and and rechecked that all Lua files in `src` report no errors with luacheck. The tests are passing now. Once again sorry for the tardiness on my part.

blueyed commented:
Also this appears to have been rebased/merged wrongly in some way.

blueyed commented:
> Once again sorry for the tardiness on my part.

No worries.  After all Justin said to take over anyway..?

This should also run `luacheck` then with the lint job I assume.

blueyed commented on src/nvim/options.lua:
```diff
@@ -188,7 +188,7 @@ return {
-      deny_duplicates=true,
+      deny_duplicates=true, -- luacheck: ignore 314
       type='string', list='comma', scope={'global'},
       deny_duplicates=true,
```
That's an actual error that should not be ignored.

sameedali replied:
Thanks. Fixed it

sameedali commented:
> Also this appears to have been rebased/merged wrongly in some way.

Yeah, i think I made a mistake and force pushed once. I don't know how to correct that. Should I redo all changes in one commit and generate another pull request? I'm a little confused as to how should I fix this?

sameedali committed:
Added luacheckrc file in src
- Fixed a missed linting error in options.lua
- removed the luacheck ignore flags and added luacheckrc file
Changes to be committed:
      new file:   src/.luacheckrc
      modified:   src/nvim/lua/vim.lua
      modified:   src/nvim/options.lua

blueyed commented:
Force-pushing is fine, but it looks like you've rebased to much..
Try `git rebase -i origin/master` and then select/keep/pick only your commits.

blueyed commented:
@sameedali 
Also just for information: it is usually a good idea to use a separate Git branch for PRs, but this cannot be changed now - just remember it for future submissions.

sameedali committed:
[skip appveyor] Fixed error after rebase

sameedali commented:
> Force-pushing is fine, but it looks like you've rebased to much..
> Try `git rebase -i origin/master` and then select/keep/pick only your commits.

@blueyed I just rebased it. I think it is correct now

blueyed commented:
@sameedali 
Looks better, thanks!

As for cmake/make targets:
In general I think we should have a `lint-lua` target then, which gets called/used by `lint` then also, and the existing `lint` might be renamed/factored out to `lint-clint` then.
But let's wait for @justinmk / others for this.

bfredl commented on src/nvim/lua/vim.lua:
```diff
@@ -38,8 +38,8 @@
 local function _system(cmd)
-  local out = vim.api.nvim_call_function('system', { cmd })
-  local err = vim.api.nvim_get_vvar('shell_error')
+  local out = vim.api.nvim_call_function('system', { cmd }) --luacheck: ignore 113
```
can these be removed, now that the luacheckrc is added?

blueyed replied:
Yes.

sameedali replied:
I just removed them.

sameedali committed:
[skip appveyor] Fixed linting errors in vim.lua
After rebasing the src/nvim/vim.lua contained the redundant luacheck ignore flags.
This commit removes them.

blueyed commented:
Note/JFI: there's `test/.luacheckrc` already.

bfredl commented:
LGTM, is this ready? Perhaps need a rebase to check with recent lua changes.

blueyed commented:
Finishing this in https://github.com/neovim/neovim/pull/10414.
